### PR TITLE
feat: add tag as tuples through setter

### PR DESCRIPTION
**Pull request by AdolfoVillalobos**
This PR addresses https://github.com/pydantic/logfire/issues/477.

From now on, the property LogfireSpan.tags will return a tuple and can be modified using common tuple operations.

Trying to use list methods will raise errors from now on.

This is a first attempt. I'm happy to improve and receive feedback.

### CODE REVIEW

File: logfire/_internal/main.py [hunk 1]
```diff
@@ -1698,7 +1698,17 @@ def message_template(self) -> str | None:  # pragma: no cover
 
     @property
     def tags(self) -> Sequence[str]:  # pragma: no cover
```
Comment (alexmojaki) [inline logfire/_internal/main.py]:
```suggestion
    def tags(self) -> tuple[str, ...]:
```

Please test this case.

File: logfire/_internal/main.py [hunk 2]
```diff
@@ -1698,7 +1698,17 @@ def message_template(self) -> str | None:  # pragma: no cover
...
-        return self._get_attribute(ATTRIBUTES_TAGS_KEY, [])
+        return self._get_attribute(ATTRIBUTES_TAGS_KEY, ())
+
+    @tags.setter
+    @handle_internal_errors()
+    def tags(self, new_tags: Sequence[str]) -> None:
+        """Set or add tags to the span."""
+        updated_tags = uniquify_sequence(list(new_tags))
```
Comment (alexmojaki) [inline logfire/_internal/main.py]:
```suggestion
        updated_tags: tuple = uniquify_sequence(list(new_tags))
```

Comment (alexmojaki) [reply logfire/_internal/main.py]: Why the conversion to a list first?

Comment (alexmojaki) [reply logfire/_internal/main.py]: Also please test setting a non-list

File: logfire/_internal/main.py [hunk 3]
```diff
@@ -1698,7 +1698,17 @@ def message_template(self) -> str | None:  # pragma: no cover
...
+        if self._span is None:  # pragma: no cover
```
Comment (alexmojaki) [inline logfire/_internal/main.py]:
```suggestion
        if self._span is None:
```

Please test this case by setting tags before `with span:`

**Comment by AdolfoVillalobos** [comment]
@alexmojaki Thanks for the valuable feedback! My bad using "list" instead of "tuple".  

I also included a test case for setting tags when the `_span` attribute is `None`. 

I'm open to any other suggestions or improvements you might have. Thanks again!

File: logfire/_internal/main.py [hunk 4]
```diff
@@ -1697,8 +1697,18 @@ def message_template(self) -> str | None:  # pragma: no cover
...
-    def tags(self) -> Sequence[str]:  # pragma: no cover
-        return self._get_attribute(ATTRIBUTES_TAGS_KEY, [])
+    def tags(self) -> tuple[str, ...]:
+        return self._get_attribute(ATTRIBUTES_TAGS_KEY, ())
+
+    @tags.setter
+    @handle_internal_errors()
+    def tags(self, new_tags: Sequence[str]) -> None:
+        """Set or add tags to the span."""
+        updated_tags: tuple[str, ...] = uniquify_sequence(tuple(new_tags))
```
Comment (alexmojaki) [inline logfire/_internal/main.py]:
```suggestion
        updated_tags: tuple[str, ...] = uniquify_sequence(new_tags)
```

My point about `list` before was that it shouldn't need preconverting at all.

File: logfire/_internal/main.py [hunk 5]
```diff
@@ -1697,8 +1697,18 @@ def message_template(self) -> str | None:  # pragma: no cover
         return self._get_attribute(ATTRIBUTES_MESSAGE_TEMPLATE_KEY, None)
 
     @property
-    def tags(self) -> Sequence[str]:  # pragma: no cover
-        return self._get_attribute(ATTRIBUTES_TAGS_KEY, [])
+    def tags(self) -> tuple[str, ...]:
+        return self._get_attribute(ATTRIBUTES_TAGS_KEY, ())
```
Comment (alexmojaki) [inline logfire/_internal/main.py:1730]: Please test this default value

File: tests/test_logfire.py [hunk 6]
```diff
@@ -2583,3 +2583,28 @@ def test_force_flush(exporter: TestExporter):
...
+        span.tags += ('d',)
+        assert span.tags == ('a', 'b', 'c', 'd')
+
+
+def test_tags_with_setter_and_non_initialized_span(exporter: TestExporter) -> None:
+    span = logfire.with_tags('tag1', 'tag2').span('test span')
+    assert span.tags == ('tag1', 'tag2')
+
+    span.tags = ('tag3', 'tag4')
+    assert span.tags == ('tag3', 'tag4')
```
Comment (alexmojaki) [inline tests/test_logfire.py]: please test that these remain after opening the span

File: logfire/_internal/main.py (hunk 4 above)
Comment (alexmojaki) [reply logfire/_internal/main.py]: but please test what happens when setting tags to a non-tuple value

Comment (alexmojaki) [reply logfire/_internal/main.py]: also we probably need to special-case setting to a string, e.g. `span.tags = "bar"` should behave like `span.tags = ["bar"]`

File: logfire/_internal/main.py [hunk 7]
```diff
@@ -1697,8 +1697,18 @@ def message_template(self) -> str | None:  # pragma: no cover
...
+
+    @tags.setter
+    @handle_internal_errors()
+    def tags(self, new_tags: Sequence[str]) -> None:
+        """Set or add tags to the span."""
+        updated_tags: tuple[str, ...] = uniquify_sequence(tuple(new_tags))
+        if self._span is None:
+            self._otlp_attributes[ATTRIBUTES_TAGS_KEY] = updated_tags
+        else:
+            self._span.set_attribute(ATTRIBUTES_TAGS_KEY, updated_tags)
```
Comment (alexmojaki) [inline logfire/_internal/main.py]: please dedupe this logic with the message setter, probably by adding a `_set_attribute` method

**Review by alexmojaki — CHANGES_REQUESTED** [review]

**Comment by alexmojaki** [comment]
Also please update `NoopSpan`

**Comment by alexmojaki** [comment]
and `logfire_api.LogfireSpan`

File: logfire/_internal/main.py (hunk 4 above)
Comment (AdolfoVillalobos) [reply logfire/_internal/main.py]: Makes sense, my bad. 

If I understand correctly, we should allow setting tags using tuples, lists and str.  I added tests for these 3 cases, enforcing that the output is always a tuple.

In the case of lists, the setter will fail if the try to use a concatenation operation "+=". That kind of operations will only be available for tuples. 

My question is: Should we keep it like this, or should we also allow for concatenation operations in the case of lists? 

```python
# is valid
tags += ("a", "b")

# will fail
tags += ["a", "b"]
```

File: logfire/_internal/main.py (hunk 7 above)
Comment (AdolfoVillalobos) [reply logfire/_internal/main.py]: Makes sense. I created the `_set_attribute` method.  I assume that we should not add its signature to the `logfire_api.LogfireSpan` declarations.  

Not adding it implies that pyright does not allow for "direct" testing (I think)

File: logfire/_internal/main.py [hunk 8]
```diff
@@ -173,7 +173,8 @@ def _span(
 
             tags = (self._tags or ()) + tuple(_tags or ())
             if tags:
-                otlp_attributes[ATTRIBUTES_TAGS_KEY] = uniquify_sequence(tags)
+                updated_tags = uniquify_sequence(tags)
+                otlp_attributes[ATTRIBUTES_TAGS_KEY] = updated_tags
```
Comment (alexmojaki) [inline logfire/_internal/main.py]: please revert

File: logfire/_internal/main.py [hunk 9]
```diff
@@ -1697,19 +1698,25 @@ def message_template(self) -> str | None:  # pragma: no cover
         return self._get_attribute(ATTRIBUTES_MESSAGE_TEMPLATE_KEY, None)
 
     @property
-    def tags(self) -> Sequence[str]:  # pragma: no cover
-        return self._get_attribute(ATTRIBUTES_TAGS_KEY, [])
+    def tags(self) -> tuple[str, ...]:
+        return tuple(self._get_attribute(ATTRIBUTES_TAGS_KEY, ()))
```
Comment (alexmojaki) [inline logfire/_internal/main.py]: rather store a tuple than convert on get

File: logfire/_internal/main.py [hunk 10]
```diff
@@ -1793,6 +1800,13 @@ def _get_attribute(self, key: str, default: Any) -> Any:
         attributes = getattr(self._span, 'attributes', self._otlp_attributes)
         return attributes.get(key, default)
 
+    def _set_attribute(self, key: str, value: Any) -> None:
```
Comment (alexmojaki) [inline logfire/_internal/main.py:1830]:
```suggestion
    def _set_attribute(self, key: str, value: otel_types.AttributeValue) -> None:
```

File: logfire-api/logfire_api/_internal/main.pyi [hunk 11]
```diff
@@ -932,7 +932,9 @@ class LogfireSpan(ReadableSpan):
     @property
     def message_template(self) -> str | None: ...
     @property
-    def tags(self) -> Sequence[str]: ...
+    def tags(self) -> tuple[str, ...]: ...
+    @tags.setter
+    def tags(self, new_tags: Sequence[str]) -> None: ...
```
Comment (alexmojaki) [inline logfire-api/logfire_api/_internal/main.pyi:937]: The actual class in `logfire_api/__init__.py` needs updating so that setting tags does nothing rather than raising an error, and this needs testing.

File: logfire/_internal/main.py [hunk 12]
```diff
@@ -1828,8 +1842,8 @@ def message_template(self) -> str:  # pragma: no cover
         return ''
 
     @property
-    def tags(self) -> Sequence[str]:  # pragma: no cover
```
Comment (alexmojaki) [inline logfire/_internal/main.py:1860]: See the comment on `@message.setter`, tags have the same need, and testing in `test_internal_exception_span`

File: tests/test_logfire.py [hunk 13]
```diff
@@ -1548,50 +1548,6 @@ def test_readable_span_signature():
     }
 
 
-def test_tags(exporter: TestExporter) -> None:
```
Comment (alexmojaki) [inline tests/test_logfire.py]: Why was this test moved? If something changed, I can't see what.

File: tests/test_logfire.py [hunk 14]
```diff
@@ -2583,3 +2539,109 @@ def test_force_flush(exporter: TestExporter):
...
+    # Call __enter__ to initialize _span, check that tags remain unchanged
+    with span:
+        assert span.tags == ('tag3', 'tag4')
+
+    # Check that tags remain unchanged
+    assert span.tags == ('tag3', 'tag4')
+
+
+def test_tags_setter_valid_type_extensions() -> None:
+    span = logfire.span('test span')
```
Comment (alexmojaki) [inline tests/test_logfire.py]:
```suggestion
    span = logfire.span('test span')
    assert span.tags == ()
```

**Review by alexmojaki — CHANGES_REQUESTED** [review]

File: tests/test_logfire.py (hunk 13 above)
Comment (AdolfoVillalobos) [reply tests/test_logfire.py]: I moved it so that all the tag tests are "near each other." If that's an issue, I can revert it.

File: tests/test_logfire.py (hunk 14 above)
Comment (AdolfoVillalobos) [reply tests/test_logfire.py]: I added tests checking default values in many cases.

File: logfire-api/logfire_api/_internal/main.pyi (hunk 11 above)
Comment (AdolfoVillalobos) [reply logfire-api/logfire_api/_internal/main.pyi:937]: My bad! I've now updated the `__init__.py` and `__init__.pyi` files to account for the new setter and signature changes.

File: logfire/_internal/main.py (hunk 12 above)
Comment (AdolfoVillalobos) [reply logfire/_internal/main.py:1860]: I've added:

```python
    # This is required to make `span.tags = ` not raise an error.
    @tags.setter
    def tags(self, new_tags: Sequence[str]) -> None:
        pass
       
```
        
on the `NoopSpan` class so that setting `span.tags = "bar"` does not raise errors and does nothing. I hope I understood correctly.

File: tests/test_logfire.py (hunk 13 above)
Comment (alexmojaki) [reply tests/test_logfire.py]: let's keep diffs minimal and clean, both to make this PR more readable and for future `git blame`s. i don't think it's very important for the tests to be near each other, but you can also move the new tests here.

**Review by alexmojaki — APPROVED** [review]
Thank you!
